lenskit-solver/src/main/java/org/lenskit/featurizer/FeatureExtractor.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * LensKit, an open source recommender systems toolkit.
+ * Copyright 2010-2014 LensKit Contributors.  See CONTRIBUTORS.md.
+ * Work on LensKit has been funded by the National Science Foundation under
+ * grants IIS 05-34939, 08-08692, 08-12148, and 10-17697.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+package org.lenskit.featurizer;
+
+import org.lenskit.data.entities.Entity;
+import org.lenskit.space.IndexSpace;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The general interface for feature extractors.
+ *
+ * @author <a href="http://www.grouplens.org">GroupLens Research</a>
+ */
+public interface FeatureExtractor extends Serializable {
+
+    /**
+     * Feature extractor interface.
+     *
+     * @param entity the data entity to extract feature from.
+     * @param indexSpace the look-up index space to use
+     * @param update whether update indexSpace if the used attributes are not present in the indexSpace
+     * @return a feature map from feature names to a list of features.
+     */
+    Map<String, List<Feature>> extract(Entity entity, boolean update,
+                                       IndexSpace indexSpace);
```

Comment:
I'm unconvinced about allowing index spaces to be modified. Consider an immutable structure.